src/response/metrics_hierarchy.rs

```diff
@@ -0,0 +1,107 @@
+use std::prelude::v1::*;
+
+use std::{
+    collections::{
+        btree_map::{Iter, IterMut},
+        BTreeMap,
+    },
+    fmt,
+};
+
+use serde::{
+    de::{self, Deserializer, MapAccess, Visitor},
+    Deserialize,
+};
+
+// We might want to consider moving from a BTreeMap to a Vec, as most of the time this btreemap will
+// contain a (very) small number of entries.
+#[derive(Debug, Clone, PartialEq, Default)]
+pub struct MetricsHierarchy(BTreeMap<String, Vec<String>>);
+
+impl MetricsHierarchy {
+    pub fn new() -> Self {
+        Self(BTreeMap::new())
+    }
+
+    pub fn insert<S: Into<String>, V: Into<Vec<String>>>(
+        &mut self,
+        parent_metric: S,
+        children_metrics: V,
+    ) -> Option<Vec<String>> {
+        self.0.insert(parent_metric.into(), children_metrics.into())
+    }
+
+    pub fn remove<S: AsRef<str>>(&mut self, parent_metric: S) -> Option<Vec<String>> {
+        self.0.remove(parent_metric.as_ref())
+    }
+
+    pub fn iter(&self) -> Iter<String, Vec<String>> {
+        self.0.iter()
+    }
+
+    pub fn iter_mut(&mut self) -> IterMut<String, Vec<String>> {
+        self.0.iter_mut()
+    }
+
+    pub fn into_inner(self) -> BTreeMap<String, Vec<String>> {
+        self.0
+    }
+
+    /// Retrieves the parent metric of a given metric. Note that Apisonator metrics have 0 or
+    /// 1 parent metrics, not multiple.
+    pub fn parent_of(&self, metric_name: &str) -> Option<&str> {
+        self.iter().find_map(|(parent, v)| {
```

Comment:
This doesn't look very efficient, but I guess it's not going to be a problem for the number of metrics that we can expect in a response.

Reply:
This is not added by this PR, just moved code, but regarding efficiency for large numbers of metrics, then I'd just call `into_inner()` and use the `BTreeMap` directly to populate a similar map based on the `parent_of` relationship for fast queries. It's still something you might only want to do in those occasions and you still can do that as is - the only issue I'd see is having an estimate as to how big the map is, for which we might lack a good way, ie. maybe `.iter().count()` can be optimized, maybe not, so in that case we might want to have an `fn as_map(&self) -> &BTreeMap` method...

But that's flour of another costal.